upgrade/scripts/upgrade/rollout-restart.sh

```diff
@@ -0,0 +1,259 @@
+    return 0 # Pod has the latest Istio versions
+  else
+    return 1 # Pod does not have the latest Istio versions
+  fi
+}
+
+# Function to determine the controlling resource of a pod
+get_controlling_resource() {
+  local namespace=$1
+  local pod=$2
+  owner_references=$(kubectl get pod "$pod" -n "$namespace" -o jsonpath="{.metadata.ownerReferences[0].kind}/{.metadata.ownerReferences[0].name}")
+
+  if [[ -n $owner_references ]]; then
+    echo "$owner_references"
+  else
+    # Fallback to describe to get owner information if needed
+    controlling_resource=$(kubectl describe pod "$pod" -n "$namespace" | grep -E "Controlled By" | awk -F: '{print $2}' | xargs)
+    echo "$controlling_resource"
+  fi
+}
+
+# Function to check if a pod uses the Istio proxy
+check_pod_uses_istio_proxy() {
+  local namespace=$1
+  local pod=$2
+  local containers
+  containers=$(kubectl get pod "$pod" -n "$namespace" -o jsonpath="{.spec.containers[*].name}")
+
+  if echo "$containers" | grep -q "istio-proxy"; then
+    return 0 # Pod uses Istio proxy
+  else
+    return 1 # Pod does not use Istio proxy
+  fi
+}
+
+# Function to perform rollout restart and check status for a given resource
+restart_and_check_status() {
+  local namespace=$1
+  shift
+  local resources=("$@")
+
+  for resource in "${resources[@]}"; do
+    resource_type=$(echo "$resource" | cut -d'/' -f1)
+    resource_name=$(echo "$resource" | cut -d'/' -f2)
+
+    if [[ $resource_type == "ReplicaSet" ]]; then
+      # Find the corresponding Deployment
+      deployment=$(kubectl get replicasets "$resource_name" -n "$namespace" -o jsonpath="{.metadata.ownerReferences[0].name}")
+      if [[ -n $deployment ]]; then
+        resource_type="Deployment"
+        resource_name=$deployment
+      else
+        echo "No corresponding Deployment found for ReplicaSet $resource_name"
+        continue
+      fi
+    fi
+
+    echo "Rolling out restart for $resource_type/$resource_name in namespace: $namespace"
+    kubectl rollout restart "$resource_type"/"$resource_name" -n "$namespace"
+  done
+
+  echo "Waiting for 3 minutes..."
+  sleep 180
+
+  for resource in "${resources[@]}"; do
+    resource_type=$(echo "$resource" | cut -d'/' -f1)
+    resource_name=$(echo "$resource" | cut -d'/' -f2)
+
+    if [[ $resource_type == "ReplicaSet" ]]; then
+      # Find the corresponding Deployment
+      deployment=$(kubectl get replicasets "$resource_name" -n "$namespace" -o jsonpath="{.metadata.ownerReferences[0].name}")
+      if [[ -n $deployment ]]; then
+        resource_type="Deployment"
+        resource_name=$deployment
+      else
+        echo "No corresponding Deployment found for ReplicaSet $resource_name"
+        continue
+      fi
+    fi
+
+    echo "Checking rollout status for $resource_type/$resource_name in namespace: $namespace"
+    kubectl rollout status "$resource_type"/"$resource_name" -n "$namespace"
+  done
```

Comment:
These lines are identical to lines 158-176.  Put them into a single function, and call the function twice instead of replicating the block of code.

Reply:
In lines 158-176 we are performing a `rollout restart` of the resources and in 181-199 we are checking the `rollout status` of the corresponding resources. 
Although the part for fetching the `Deployment` (if the resource is a `Replicaset`) is same the functionality is different. To combine both of them in a function will create scope issues Sir.

Reply:
Yes David, considering the advise we have updated the code to avoid replication of code.

Reply:
The latest commit which has been pushed has this issue addressed @dborman-hpe